### PR TITLE
OCPBUGS-27961 Procedure for configuring Whereabouts IP reconciler schedule

### PR DESCRIPTION
**Pull request by kquinn1204**
[OCPBUGS-27961]: Procedure for configuring Whereabouts IP reconciler

Version(s): 4.15, 4.14 and main

Issue:  https://issues.redhat.com/browse/OCPBUGS-27961 (OCPBUGS](https://issues.redhat.com/browse/OCPBUGS-26986)

Link to docs preview: https://71081--ocpdocs-pr.netlify.app/openshift-enterprise/latest/networking/multiple_networks/configuring-additional-network#nw-multus-configuring-whereabouts-ip-reconciler-schedule_configuring-additional-network

QE review:
- [x] QE has approved this change.

Additional information:
See PR https://github.com/openshift/openshift-docs/pull/70818.

### CODE REVIEW

**Comment by kquinn1204** [comment]
@cgoncalves @liornoy please add LGTM if happy with changes

File: modules/nw-multus-ipam-object.adoc [hunk 1]
```diff
@@ -310,4 +310,72 @@ pod/whereabouts-reconciler-p4sxw 1/1 Running 0 6s
 pod/whereabouts-reconciler-rvfdv 1/1 Running 0 6s
 pod/whereabouts-reconciler-svzw9 1/1 Running 0 6s
 daemonset.apps/whereabouts-reconciler 6 6 6 6 6 kubernetes.io/os=linux 6s
+----
+
+[id="nw-multus-configuring-whereabouts-ip-reconciler-schedule_{context}"]
+== Configuring the Whereabouts IP reconciler schedule
+
+The Whereabouts IPAM CNI plugin runs the IP reconciler daily in order to clean up any stranded IP allocations that may result in exhausting IPs and hence prevent new Pods from getting an IP allocated to them.
```
Comment (cbippley) [inline modules/nw-multus-ipam-object.adoc:318]:
```suggestion
The Whereabouts IPAM CNI plugin runs the IP reconciler daily to clean up any stranded IP allocations that may result in exhausting IPs and therefore prevent new Pods from getting an IP allocated to them.
```
Vale is suggesting 'therefore' instead of 'hence' but I can't find where that lives in the style guidance. Also, the 'exhausting IPs' makes the sentence a little confusing, maybe rephrase to be more clear?

File: modules/nw-multus-ipam-object.adoc [hunk 2]
```diff
@@ -310,4 +310,72 @@ pod/whereabouts-reconciler-p4sxw 1/1 Running 0 6s
...
+.Procedure
+
+. Run the following command to create a `ConfigMap` named `whereabouts-config` in the `openshift-multus` namespace with a specific cron expression for the IP reconciler:
++
+[source,terminal]
+----
+$ oc create configmap whereabouts-config -n openshift-multus --from-literal=reconciler_cron_expression="*/15 * * * *"
+----
++
+This cron expression indicates the IP reconciler should run every 15 minutes. Adjust the expression based on your specific requirements.
```
Comment (cbippley) [inline modules/nw-multus-ipam-object.adoc:335]:
```suggestion
This cron expression indicates the IP reconciler runs every 15 minutes. Adjust the expression based on your specific requirements. 
```
Does it run reliably every 15 minutes or not? Maybe add something about if it doesn't run on schedule if there is a chance it won't?

File: modules/nw-multus-ipam-object.adoc [hunk 3]
```diff
@@ -310,4 +310,72 @@ pod/whereabouts-reconciler-p4sxw 1/1 Running 0 6s
...
+pod/whereabouts-reconciler-2p7hw                   1/1     Running   0             4m14s
+pod/whereabouts-reconciler-76jk7                   1/1     Running   0             4m14s
+pod/whereabouts-reconciler-94zw6                   1/1     Running   0             4m14s
+pod/whereabouts-reconciler-mfh68                   1/1     Running   0             4m14s
+pod/whereabouts-reconciler-pgshz                   1/1     Running   0             4m14s
+pod/whereabouts-reconciler-xn5xz                   1/1     Running   0             4m14s
+daemonset.apps/whereabouts-reconciler          6         6         6       6            6           kubernetes.io/os=linux   4m16s
+----
+
+. Run the following command to verify that the `whereabouts-reconciler`  runs the IP reconciler with the configured interval:
```
Comment (cbippley) [inline modules/nw-multus-ipam-object.adoc:357]:
```suggestion
. Run the following command to verify that the `whereabouts-reconciler` runs the IP reconciler with the configured interval:
```

**Review by cbippley — COMMENTED** [review]
A couple of suggestions. Don't forget to get an ACK from QE before asking for a merge review. Great job!

File: modules/nw-multus-ipam-object.adoc [hunk 4]
```diff
@@ -310,4 +310,72 @@ pod/whereabouts-reconciler-p4sxw 1/1 Running 0 6s
...
+[id="nw-multus-configuring-whereabouts-ip-reconciler-schedule_{context}"]
+== Configuring the Whereabouts IP reconciler schedule
+
+The Whereabouts IPAM CNI plugin runs the IP reconciler daily to clean up any stranded IP allocations that may result in exhausting IPs and therefore prevent new Pods from getting an IP allocated to them.
+
+The following procedure shows how to change the frequency at which the IP reconciler runs. 
+
+.Prerequisites
+* You installed the OpenShift CLI (`oc`).
+* You have access to the cluster as a user with the `cluster-admin` role.
```
Comment (liornoy) [inline modules/nw-multus-ipam-object.adoc:324]: Sorry for the late response, I just thought about it now while re-reviewing this just to make sure.
I believe we should add here in the prerequisites the fact that the whereabouts daemonset should be deployed, with the `whereabouts-reconciler` pods up and running.
That's because that's where 'magic' happens and without it, the presence of the ConfigMap 
would do nothing.

File: modules/nw-multus-ipam-object.adoc [hunk 5]
```diff
@@ -310,4 +310,72 @@ pod/whereabouts-reconciler-p4sxw 1/1 Running 0 6s
...
+.Procedure
+
+. Run the following command to create a `ConfigMap` named `whereabouts-config` in the `openshift-multus` namespace with a specific cron expression for the IP reconciler:
++
+[source,terminal]
+----
+$ oc create configmap whereabouts-config -n openshift-multus --from-literal=reconciler_cron_expression="*/15 * * * *"
+----
++
+This cron expression indicates the IP reconciler runs every 15 minutes. Adjust the expression based on your specific requirements.
```
Comment (liornoy) [inline modules/nw-multus-ipam-object.adoc:335]: Also, here - regarding "Adjust the expression based on your specific requirements":
It's important to mention the **limitation** of the whereabouts reconciler that can only consume a cron expression pattern that includes 5 asterisks and not 6.
The 6th would be used for Seconds, **but it is currently not supported.**

Comment (liornoy) [reply modules/nw-multus-ipam-object.adoc:335]: e.g. a customer may try to set: `*/30 * * * * *`
so it will run every 30 sec, but that will cause the reconciler to err out
parsing the expression, and revert to the default once a day setting

File: modules/nw-multus-ipam-object.adoc [hunk 6]
```diff
@@ -310,4 +310,78 @@ pod/whereabouts-reconciler-p4sxw 1/1 Running 0 6s
...
+== Configuring the Whereabouts IP reconciler schedule
+
+The Whereabouts IPAM CNI plugin runs the IP reconciler daily to clean up any stranded IP allocations that may result in exhausting IPs and therefore prevent new Pods from getting an IP allocated to them.
+
+The following procedure shows how to change the frequency at which the IP reconciler runs. 
+
+.Prerequisites
+* You installed the OpenShift CLI (`oc`).
+* You have access to the cluster as a user with the `cluster-admin` role.
+* You have deployed the `whereabouts-reconciler` daemon set, and the `whereabouts-reconciler` pods are up and running.
```
Comment (liornoy) [inline modules/nw-multus-ipam-object.adoc:325]: That's great, is there a place to reference the relevant docs?
https://docs.openshift.com/container-platform/4.14/networking/multiple_networks/configuring-additional-network.html#nw-multus-creating-whereabouts-reconciler-daemon-set_configuring-additional-network

**Review by liornoy — COMMENTED** [review]
Just a small suggestion, other than that, LGTM, thanks

File: modules/nw-multus-ipam-object.adoc (hunk 6 above)
Comment (kquinn1204) [reply modules/nw-multus-ipam-object.adoc:325]: We can't add a cross ref in the module and this section is just before this so I think we are OK. TBH when i followed the guidance in that section "Modify the additionalNetworks parameter in the CR to add the whereabouts-shim network attachment definition." the additionalNetworks section did not exist and I needed to create it. I will update in separate PR.

File: modules/nw-multus-ipam-object.adoc [hunk 7]
```diff
@@ -310,4 +310,78 @@ pod/whereabouts-reconciler-p4sxw 1/1 Running 0 6s
...
 pod/whereabouts-reconciler-svzw9 1/1 Running 0 6s
 daemonset.apps/whereabouts-reconciler 6 6 6 6 6 kubernetes.io/os=linux 6s
+----
+
+[id="nw-multus-configuring-whereabouts-ip-reconciler-schedule_{context}"]
+== Configuring the Whereabouts IP reconciler schedule
+
+The Whereabouts IPAM CNI plugin runs the IP reconciler daily to clean up any stranded IP allocations that may result in exhausting IPs and therefore prevent new Pods from getting an IP allocated to them.
+
+The following procedure shows how to change the frequency at which the IP reconciler runs.
```
Comment (JoeAldinger) [inline modules/nw-multus-ipam-object.adoc:320]:
```suggestion
The following procedure shows how to change the frequency at which the IP reconciler runs. 
```
Try to avoid self referential [language](Avoid self-referential language, such as "This topic covers…​" or "Use this procedure to…​".

).

File: modules/nw-multus-ipam-object.adoc [hunk 8]
```diff
@@ -310,4 +310,78 @@ pod/whereabouts-reconciler-p4sxw 1/1 Running 0 6s
 pod/whereabouts-reconciler-rvfdv 1/1 Running 0 6s
 pod/whereabouts-reconciler-svzw9 1/1 Running 0 6s
 daemonset.apps/whereabouts-reconciler 6 6 6 6 6 kubernetes.io/os=linux 6s
+----
+
+[id="nw-multus-configuring-whereabouts-ip-reconciler-schedule_{context}"]
+== Configuring the Whereabouts IP reconciler schedule
```
Comment (JoeAldinger) [inline modules/nw-multus-ipam-object.adoc:316]: This a concept module that two procedures in it, which is not good. I think these two procedures need to be refractor into separate modules and then added to the appropriate assemblies.

https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#writing-procedures

**Review by JoeAldinger — COMMENTED** [review]
This procedure and the one above it need to be refactored into separate modules please.

**Comment by kquinn1204** [comment]
HI @JoeAldinger This should be OK now thanks.

File: modules/nw-multus-creating-whereabouts-reconciler-daemon-set.adoc [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+:_mod-docs-content-type: PROCEDURE
+
+[id="nw-multus-creating-whereabouts-reconciler-daemon-set_{context}"]
+= Creating a Whereabouts reconciler daemon set
+
+The Whereabouts reconciler is responsible for managing dynamic IP address assignments for the pods within a cluster using the Whereabouts IP Address Management (IPAM) solution. It ensures that each pods gets a unique IP address from the specified IP address range. It also handles IP address releases when pods are deleted or scaled down.
+
+[NOTE]
+====
+You can also use a `NetworkAttachmentDefinition` custom resource for dynamic IP address assignment.
```
Comment (skopacz1) [inline modules/nw-multus-creating-whereabouts-reconciler-daemon-set.adoc:14]: Should this have (CR)?
```suggestion
You can also use a `NetworkAttachmentDefinition` custom resource (CR) for dynamic IP address assignment.
```

File: modules/nw-multus-creating-whereabouts-reconciler-daemon-set.adoc [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+====
+
+The Whereabouts reconciler daemon set is automatically created when you configure an additional network through the Cluster Network Operator. It is not automatically created when you configure an additional network from a YAML manifest.
+
+To trigger the deployment of the Whereabouts reconciler daemon set, you must manually create a `whereabouts-shim` network attachment by editing the Cluster Network Operator custom resource file.
```
Comment (skopacz1) [inline modules/nw-multus-creating-whereabouts-reconciler-daemon-set.adoc:19]: Same question as above:
```suggestion
To trigger the deployment of the Whereabouts reconciler daemon set, you must manually create a `whereabouts-shim` network attachment by editing the Cluster Network Operator custom resource (CR) file.
```

**Review by skopacz1 — COMMENTED** [review]
Left a few comments but otherwise looks good! Great stuff!

File: modules/nw-multus-configuring-whereabouts-ip-reconciler-schedule.adoc [hunk 11]
```diff
@@ -0,0 +1,79 @@
+// Module included in the following assemblies:
+//
+// * networking/multiple_networks/configuring-additional-network.adoc
+
+:_mod-docs-content-type: PROCEDURE
+
+[id="nw-multus-configuring-whereabouts-ip-reconciler-schedule_{context}"]
+= Configuring the Whereabouts IP reconciler schedule
+
+The Whereabouts IPAM CNI plugin runs the IP reconciler daily to clean up any stranded IP allocations that may result in exhausting IPs and therefore prevent new Pods from getting an IP allocated to them.
```
[the remark this answers is not in the document]
Comment (kquinn1204) [reply modules/nw-multus-configuring-whereabouts-ip-reconciler-schedule.adoc:10]: I believe plugin is now the accepted way

File: modules/nw-multus-configuring-whereabouts-ip-reconciler-schedule.adoc [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
++
+[source,terminal]
+----
+pod/whereabouts-reconciler-2p7hw                   1/1     Running   0             4m14s
+pod/whereabouts-reconciler-76jk7                   1/1     Running   0             4m14s
+pod/whereabouts-reconciler-94zw6                   1/1     Running   0             4m14s
+pod/whereabouts-reconciler-mfh68                   1/1     Running   0             4m14s
+pod/whereabouts-reconciler-pgshz                   1/1     Running   0             4m14s
+pod/whereabouts-reconciler-xn5xz                   1/1     Running   0             4m14s
+daemonset.apps/whereabouts-reconciler          6         6         6       6            6           kubernetes.io/os=linux   4m16s
```
[the remark this answers is not in the document]
Comment (kquinn1204) [reply modules/nw-multus-configuring-whereabouts-ip-reconciler-schedule.adoc:52]: this is the actual output and it looks that way in the output

File: modules/nw-multus-configuring-whereabouts-ip-reconciler-schedule.adoc (hunk 11 above)
[the remark this answers is not in the document]
Comment (skopacz1) [reply modules/nw-multus-configuring-whereabouts-ip-reconciler-schedule.adoc]: I didn't realize that the RHSSG's word usage section has the exact _opposite_ ruling as the ISG on this, how annoying haha. But thanks for letting me know!

File: modules/nw-multus-configuring-whereabouts-ip-reconciler-schedule.adoc [hunk 13]
```diff
@@ -0,0 +1,79 @@
+// Module included in the following assemblies:
+//
+// * networking/multiple_networks/configuring-additional-network.adoc
+
+:_mod-docs-content-type: PROCEDURE
+
+[id="nw-multus-configuring-whereabouts-ip-reconciler-schedule_{context}"]
+= Configuring the Whereabouts IP reconciler schedule
+
+The Whereabouts IPAM CNI plugin runs the IP reconciler daily to clean up any stranded IP allocations that might result in exhausting IPs and therefore prevent new Pods from getting an IP allocated to them.
```
Comment (adellape) [inline modules/nw-multus-configuring-whereabouts-ip-reconciler-schedule.adoc:10]: Lowercase no-markup for [general reference](https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#general-references):
```suggestion
The Whereabouts IPAM CNI plugin runs the IP reconciler daily to clean up any stranded IP allocations that might result in exhausting IPs and therefore prevent new pods from getting an IP allocated to them.
```
Also seems a bit long of a sentence, but it does seem tricky to break up. Not critical, but, a crack at it:
```suggestion
The Whereabouts IPAM CNI plugin runs the IP reconciler daily to clean up any stranded IP allocations. Otherwise, IP pools could become exhausted and therefore prevent new pods from getting an IP allocated to them.
```
Not sure that that's any better... 😅

File: modules/nw-multus-configuring-whereabouts-ip-reconciler-schedule.adoc [hunk 14]
```diff
@@ -0,0 +1,79 @@
...
+Use this procedure to change the frequency at which the IP reconciler runs. 
+
+.Prerequisites
+* You installed the OpenShift CLI (`oc`).
+* You have access to the cluster as a user with the `cluster-admin` role.
+* You have deployed the `whereabouts-reconciler` daemon set, and the `whereabouts-reconciler` pods are up and running.
+
+.Procedure
+
+. Run the following command to create a `ConfigMap` named `whereabouts-config` in the `openshift-multus` namespace with a specific cron expression for the IP reconciler:
```
Comment (adellape) [inline modules/nw-multus-configuring-whereabouts-ip-reconciler-schedule.adoc:21]:
```suggestion
. Run the following command to create a `ConfigMap` object named `whereabouts-config` in the `openshift-multus` namespace with a specific cron expression for the IP reconciler:
```

File: modules/nw-multus-configuring-whereabouts-ip-reconciler-schedule.adoc [hunk 15]
```diff
@@ -0,0 +1,79 @@
...
+[source,terminal]
+----
+$ oc create configmap whereabouts-config -n openshift-multus --from-literal=reconciler_cron_expression="*/15 * * * *"
+----
++
+This cron expression indicates the IP reconciler runs every 15 minutes. Adjust the expression based on your specific requirements.
++
+[NOTE]
+====
+The `whereabouts-reconciler` can only consume a cron expression pattern that includes five asterisks. The sixth, which is used to denote seconds, is currently not supported.
```
Comment (adellape) [inline modules/nw-multus-configuring-whereabouts-ip-reconciler-schedule.adoc:32]: Is there something more specific to call this here?
```suggestion
The `whereabouts-reconciler` daemon set can only consume a cron expression pattern that includes five asterisks. The sixth, which is used to denote seconds, is currently not supported.
```
Edit after making a few other related comments: You could also switch these naked `whereabouts-reconciler` instances to more generally "the Whereabouts reconciler", which I see is used in the other procedure.

File: modules/nw-multus-configuring-whereabouts-ip-reconciler-schedule.adoc [hunk 16]
```diff
@@ -0,0 +1,79 @@
...
+====
+
+. Retrieve information about resources related to the `whereabouts-reconciler` within the `openshift-multus` namespace by running the following command:
```
Comment (adellape) [inline modules/nw-multus-configuring-whereabouts-ip-reconciler-schedule.adoc:35]:
```suggestion
. Retrieve information about resources related to the `whereabouts-reconciler` daemon set and pods within the `openshift-multus` namespace by running the following command:
```
?

File: modules/nw-multus-configuring-whereabouts-ip-reconciler-schedule.adoc [hunk 17]
```diff
@@ -0,0 +1,79 @@
...
+pod/whereabouts-reconciler-2p7hw                   1/1     Running   0             4m14s
+pod/whereabouts-reconciler-76jk7                   1/1     Running   0             4m14s
+pod/whereabouts-reconciler-94zw6                   1/1     Running   0             4m14s
+pod/whereabouts-reconciler-mfh68                   1/1     Running   0             4m14s
+pod/whereabouts-reconciler-pgshz                   1/1     Running   0             4m14s
+pod/whereabouts-reconciler-xn5xz                   1/1     Running   0             4m14s
+daemonset.apps/whereabouts-reconciler          6         6         6       6            6           kubernetes.io/os=linux   4m16s
+----
+
+. Run the following command to verify that the `whereabouts-reconciler` runs the IP reconciler with the configured interval:
```
Comment (adellape) [inline modules/nw-multus-configuring-whereabouts-ip-reconciler-schedule.adoc:55]:
```suggestion
. Run the following command to verify that the `whereabouts-reconciler` pod runs the IP reconciler with the configured interval:
```

File: modules/nw-multus-creating-whereabouts-reconciler-daemon-set.adoc [hunk 18]
```diff
@@ -0,0 +1,76 @@
+// Module included in the following assemblies:
+//
+// * networking/multiple_networks/configuring-additional-network.adoc
+
+:_mod-docs-content-type: PROCEDURE
+
+[id="nw-multus-creating-whereabouts-reconciler-daemon-set_{context}"]
+= Creating a Whereabouts reconciler daemon set
+
+The Whereabouts reconciler is responsible for managing dynamic IP address assignments for the pods within a cluster using the Whereabouts IP Address Management (IPAM) solution. It ensures that each pods gets a unique IP address from the specified IP address range. It also handles IP address releases when pods are deleted or scaled down.
```
Comment (adellape) [inline modules/nw-multus-creating-whereabouts-reconciler-daemon-set.adoc:10]:
```suggestion
The Whereabouts reconciler is responsible for managing dynamic IP address assignments for the pods within a cluster using the Whereabouts IP Address Management (IPAM) solution. It ensures that each pod gets a unique IP address from the specified IP address range. It also handles IP address releases when pods are deleted or scaled down.
```

File: modules/nw-multus-configuring-whereabouts-ip-reconciler-schedule.adoc [hunk 19]
```diff
@@ -0,0 +1,79 @@
...
+= Configuring the Whereabouts IP reconciler schedule
+
+The Whereabouts IPAM CNI plugin runs the IP reconciler daily to clean up any stranded IP allocations that might result in exhausting IPs and therefore prevent new Pods from getting an IP allocated to them.
+
+Use this procedure to change the frequency at which the IP reconciler runs. 
+
+.Prerequisites
+* You installed the OpenShift CLI (`oc`).
+* You have access to the cluster as a user with the `cluster-admin` role.
+* You have deployed the `whereabouts-reconciler` daemon set, and the `whereabouts-reconciler` pods are up and running.
```
Comment (adellape) [inline modules/nw-multus-configuring-whereabouts-ip-reconciler-schedule.adoc:17]: (Now that I've seen the other procedure) Seems like this could match the style of "the Whereabouts reconciler daemon set" used in the other procedure.
```suggestion
* You have deployed the Whereabouts reconciler daemon set, and the `whereabouts-reconciler` pods are up and running.
```

Comment (kquinn1204) [reply modules/nw-multus-configuring-whereabouts-ip-reconciler-schedule.adoc:17]: looks like it should `whereabouts-reconciler`  when talking about the daemon set

**Comment by kquinn1204** [comment]
@adellape should be ready now for a merge. Thanks for your help.

**Review by adellape — APPROVED** [review]

**Comment by adellape** [comment]
Chatted further with @kquinn1204 and determined this can be backported/cherrypicked also to 4.13 (per https://access.redhat.com/errata/RHSA-2024:0660 shipping). Attempting now.
